### ISSUE TITLE
Improve ARM experience

### ISSUE THREAD

**Issue by luxas** [rancher/os#760]
This will be a tracking issue for optimizing the ARM port.

 - [x] It's working :)
 - [ ] We're using Hypriot kernel, which I think can be made smaller if we build it ourselves
 - [ ] Takes ~60 s to boot the persistent fs... I think it should be faster
 - [ ] About ~200 MB fs size when booted, maybe that could be smaller
 - [ ] Using 100MB+ RAM when booted. Why?
 - [ ] `sudo` is `-rwxr-xr-x` instead of `-rwsr-xr-x`. Can be fixed by: `chmod a=rx,u+ws /usr/bin/sudo`
 - [ ] Provide pre-built images (`.img`) at least for Raspberry Pi, which should self-resize to the SD Card on first boot
 - [ ] Document how to build it yourself

Related: #417 #735 
@imikushin @ibuildthecloud

**Comment by luxas** [comment]
Another thing: your `rancher/docker-arm` release aren't runnable on ARMv6 (RPi 1), so if we want RPi 1 support, we have to build it in with e.g. `resin/rpi-raspbian` as base image and `GOARM=6`

**Comment by DieterReuter** [comment]
@luxas If you like, we can discuss some of these points in detail to build an optimised SD card image for the RPi2. I did already built a first version, but would like to get to know your thoughts. Currently it's only running on ARMv7, thus I removed the dual kernel support and saved some additional space. 
```
wget http://downloads.hypriot.com/rancheros-rpi2-20160221.zip
```
Current build includes:
- RancherOS v0.4.3 rootfs for armhf
- single RPi2 kernel from Hypriot (no support for RPi 1!)
- clean build from scratch, not from HypriotOS
- no firmware files included (only needed to support WiFi USB dongles)

**Comment by luxas** [comment]
Yeah, maybe we should target Pi 2 first (a bird whispered that a Pi 3 is coming :smile:) and then consider a separate Pi 1 support for saving space.

Have you uploaded the build script or something? Maybe we could work on a Gist?

**Comment by luxas** [comment]
Now rancher asks for a new password on first boot. That's good. 
The `sudo` problem I had is also fixed.

Some stats on first boot: 
`free -m`: `125 MB`
`du -sh /`: `197 MB`
`ps aux | wc -l`: `76` processes
It uses pretty much `0-1%` of the cores when idle

A funny thing, when running `sudo system-docker ps`, it says that the containers were created `46 years ago` :) Probably because `1970` is the default startup time.

Will investigate even more.
I got Kubernetes running on this image now :tada: Yay!

In fact, the only thing I had to change was `sudo modprobe tun`
When running a Kubernetes master ~`640 MB` is consumed, so it's OK.

I'm excited to be a part of this work :) Let's move on to the script...

**Comment by DieterReuter** [comment]
@luxas I just used the script from @ibuildthecloud and done some optimisations. So maybe it's best, if you start with his script here https://github.com/rancher/os/tree/master/scripts/images/rpi-hypriot.

As soon as I have the time, I can cleanup my improved build script to publish it.

**Comment by spralle** [comment]
How is RPi 3 compatibility?

**Comment by DieterReuter** [comment]
@spralle You already have one? 
Honestly, we can only tell anything about the RPi 3, when we were able to test it.

**Comment by luxas** [comment]
I ordered mine ~20 mins ago. I expect it to ship in 2-3 days.
To answer your question, yes, RancherOS is RPi compatible (at least theoretically)
I'm working right now on `arm64` RancherOS rootfs though.

**Comment by spralle** [comment]
@DieterReuter, Yes but I haven't had time to play with it yet. If I can do something to test it just point me in the right direction.

**Comment by JulekFr** [comment]
Hi @DieterReuter, Thank you for the img. I am just struggling on one point, I successfully resized the partition but after reboot rootfs doesn't expand. Could you explain briefly the process?

**Comment by DieterReuter** [comment]
@JulekFr, we do have a [resizefs script](https://github.com/hypriot/image-builder-rpi/blob/master/builder/files/etc/firstboot.d/10-resize-rootdisk), which we run in HypriotOS at first boot only. The partition table of the SD card has to be changed in order to expand the rootfs. 

The same thing has to be done in RangerOS at first boot, but I think we don't have all the used Linux tools installed. So I guess, one has to modify the partition table directly and has to test this in RancherOS.

**Comment by luxas** [comment]
@JulekFr Here's the RancherOS builtin way of doing it: https://github.com/rancher/os-images/tree/master/src/docker/10-resizefs. There's a PR open for ARM support for that image too, and will hopefully be in for `v0.4.4` or something.

**Comment by luxas** [comment]
@ibuildthecloud @imikushin What are your opinions about RPi 1 support? 
I'm leaning towards that it's not worth it really, that we should see `1024MB` RAM as a hard limit.
Then we could build binaries for `ARMv7+` only, and focus on `arm64` support for RPi 3.
WDYT?

**Comment by DieterReuter** [comment]
@luxas According to RPi 3, I don't expect any improvement to run 64-bit on this board. And there is currently no support for 64-bit from the Raspberry Pi Foundation at all. The device is currently and will be limited to 1GByte for a longer time and in this case it's more efficient to use is with 32-bit userland binaries ONLY! With 1GByte you will have only disadvantages with 64-bit, the binaries will be 50% to 100% larger in size and consume more memory when running. Just my two cents.

**Comment by luxas** [comment]
Yeah, you're completely right about the RPi 3. Was just too excited about getting an ARM64 board to experiment with :smile:, so I didn't think about the 64-bit facts with larger size, RAM usage etc. compared to 32-bit ones.
Thanks for pointing that out.

The RPi is stuck at 1GB RAM, because of the VC4, IIRC?

What do you think about the Pine64 with 2GB RAM? 
Is it worth to use 64-bit there? At least it would be cool with RancherOS for ARM64
I'm still waiting for my Pine64 to ship...

**Comment by DieterReuter** [comment]
@luxas, I do have a Pine A64+ with 1GByte here for testing and right now the Linux support should be there - at least for the first tests. Maybe this would be a could experiment to compare and benchmark this board between 32- and 64-bit userland. My rough guess here: 32-bit will win.

**Comment by luxas** [comment]
I know this benchmark is quite silly, but it might give a hunch about what's better.
https://github.com/luxas/benchmark
Then one could compare if 32-bit or 64-bit compilations are faster.
WDYT?

**Comment by marcelhodan** [comment]
@DieterReuter 
Have you successfully tested the Pine A64+ ? And if so, do you still have the .img file? I would really like to add my (currently not used) Pine64 to my personal Rancher environment.

**Comment by noplanman** [comment]
Has there been any progress on getting RancherOS running on a Pine64?

**Comment by silviuvulcan** [comment]
I'm chiming in for pine64 support as well. I have a nice clusterboard (7 sopine modules, each quadcore with 2GB RAM) on which I'd like to play with rancher around.

**Comment by ppouliot** [comment]
Is there any plan for supporting server ARM64 platforms like Ampere Computing eMAG?
